jobs/kpi-forecasting/kpi_forecasting/validator.py

```diff
@@ -0,0 +1,153 @@
+        else:
+            dimension_list = []
+
+        self.input_table_full = input_table_full
+        self.dimension_list = dimension_list
+
+    def query(self) -> str:
+        """Creates a query used to generate the validation table.
+
+        Returns:
+        (str): Query to generate validation table"""
+        if len(self.dimension_list) > 0:
+            segment_comma_separated = "," + ",".join(self.dimension_list)
+        else:
+            segment_comma_separated = ""
+        # in actual_deduped, the value for historical data won't change so we can use any_value without checking forecast_trained_at
+        compare_ctes = f"""WITH forecast_deduped as (SELECT submission_date, metric_alias {segment_comma_separated},
+                                                    MAX(forecast_trained_at) as forecast_trained_at,
+                                                    ANY_VALUE(value HAVING MAX forecast_trained_at) as forecast_value 
+                                                    FROM {self.input_table_full} 
+                                                    WHERE source='forecast' AND aggregation_period='day'
+                                                    GROUP BY submission_date, metric_alias {segment_comma_separated}),
+                                actual_deduped as (SELECT submission_date, metric_alias {segment_comma_separated},
+                                                    ANY_VALUE(value) as actual_value 
+                                                    FROM {self.input_table_full}
+                                                    WHERE source='historical' AND aggregation_period='day'
+                                                    GROUP BY submission_date, metric_alias {segment_comma_separated}),
+                            compare_data as (SELECT forecast_deduped.*, actual_deduped.actual_value,
+                            (actual_deduped.actual_value-forecast_deduped.forecast_value) as difference,
+                            (actual_deduped.actual_value-forecast_deduped.forecast_value)/actual_deduped.actual_value*100 as percent_difference,
+                            ABS(actual_deduped.actual_value-forecast_deduped.forecast_value) as absolute_difference,
+                            ABS(actual_deduped.actual_value-forecast_deduped.forecast_value)/actual_deduped.actual_value*100 as absolute_percent_difference
+                                                FROM forecast_deduped INNER JOIN actual_deduped USING (submission_date, metric_alias {segment_comma_separated}))"""
+        return f"{compare_ctes} SELECT * FROM compare_data"
+
+    def fetch(self) -> pd.DataFrame:
+        """Uses the query produced by the query method to retrieve data
+            from bigquery and return as a pandas dataframe
+
+        Returns:
+            pd.DataFrame: Validation table as a pandas dataframe
+        """
+        query_job = bigquery.Client(project=self.output_project).query(self.query())
```

Comment:
I believe there's methods in the `QueryResult` (or whatever) that's returned by `Client.query` to create tables or write to tables, without needing to download the DataFrame, if there's no other data manipulation being done following the download.

Reply:
I think that's what's currently happening in the `write` method.  I left a `fetch` method because I figured we'd want a way to get the dataframe

Reply:
Ahh, duh, I didn't even look closely at the write method after reading the fetch method. My b